pandas/core/generic.py

```diff
@@ -5548,6 +5548,9 @@ def astype(
             new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors,)
             return self._constructor(new_data).__finalize__(self, method="astype")
 
+        if not results:
```

Comment:
maybe this should be in `elif is_extension_array_dtype(dtype) and self.ndim > 1:` block.

Reply:
although then the issue mentioned in https://github.com/pandas-dev/pandas/issues/33113#issuecomment-646865646 would not be fixed. maybe keep this where it is and a test for https://github.com/pandas-dev/pandas/issues/33113#issuecomment-646865646

Reply:
Let's try!

Reply:
Sorry, did not see your second comment. Ok I'll do that. Anyhow if we reach the "concat" part of the code and `results` is empty, then we have a problem.

Reply:
Although, one could argue that `pd.concat` of an empty list could return an empty frame or series instead of raising an exception.

Reply:
@simonjayhawkins Actually starting to think the other issue was not an issue but an error on the author's part. See https://github.com/pandas-dev/pandas/issues/33113#issuecomment-647176870. Awaiting your answer to proceed.

Reply:
I _think_ the other issue **is** relevant.

on master and in this PR (latest commit)...
```
>>> df = pd.DataFrame([[1, 2], [3, 4]])
>>> df
   0  1
0  1  2
1  3  4
>>>
>>> res = df.astype(dict())
>>> res
   0  1
0  1  2
1  3  4
>>>
>>> res is df
False
>>>
>>> df = pd.DataFrame([dict(), dict()])
>>>
>>> df.astype(dict())
Traceback (most recent call last):
   ...
ValueError: No objects to concatenate
>>>
```

this PR before moving https://github.com/pandas-dev/pandas/pull/33118#discussion_r443251242 ...
```
>>> df = pd.DataFrame([dict(), dict()])
>>>
>>> df.astype(dict())
Empty DataFrame
Columns: []
Index: [0, 1]
>>>
>>> res is df
False
```

so it appears astype with an empty dict is a valid op with a copy returned. so this should also work on an empty DataFrame (even if created with empty dict).

these sort of circumstances can arise on data pipelines and having consistency reduces the complexity of downstream code.

so I _think_ need to move the changes back and add a test for this.

Reply:
Ok I will revert the changes back and add a test. I still find it **very strang** to call `.astype(T)` where `T` is an instance of an object and not a type (except of course for string-defined such as `.astype("float64")`).

Reply:
DataFrame.astype can also accept a dict of column name -> data type as the dtype argument, https://pandas.pydata.org/docs/dev/reference/api/pandas.DataFrame.astype.html

Reply:
Ahhh that makes a lot of sense now.